### PR TITLE
Add an xslt test folder

### PR DESCRIPTION
**Pull request by lukebjerring**
Add a test to cover https://bugs.chromium.org/p/chromium/issues/detail?id=266305

Test cases are a translation of the internal/external cases suggested here:
https://groups.google.com/a/chromium.org/d/msg/blink-dev/XiRYbovqjkc/Ut6mzq6JBAAJ

### CODE REVIEW

**Comment by lukebjerring** [comment]
Looks like the test in its https://github.com/web-platform-tests/wpt/pull/17607/commits/6bc68e3aa63d3e20e42e516651d46a70e716f0ab state fails in all browsers:
https://wpt.fyi/results/xslt/transformToFragment.window.html?run_id=251520021&run_id=261130019&run_id=251550004

**Comment by martin-honnen** [comment]
@lukebjerring, thanks for trying to integrate the test cases. I will need some time to understand the test structure to tell why your first attempt failed. In general, for the second test with an external script file, it proved to be crucial to check the script's execution in an event listener for the `load` event. That is not related to XSLTProcessor but turned out the be necessary in a similar test with the `Range` method `createContextualFragment` (https://github.com/martin-honnen/martin-honnen.github.io/blob/master/xslt/2019/createContextualFragment-Test2.html).

**Comment by lukebjerring** [comment]
@martin-honnen - ah, my mistake. Have confirmed that both internal/external pass in firefox with https://github.com/web-platform-tests/wpt/pull/17607/commits/32c07e34c7f2c7b3f042a4fc0685064c6a36f6e5 locally.

**Comment by lukebjerring** [comment]
Also, if you haven't already stumbled across to it, see https://web-platform-tests.org/ for documentation of WPT's `testharness`

**Comment by lukebjerring** [comment]
https://github.com/web-platform-tests/wpt/pull/17607/commits/32c07e34c7f2c7b3f042a4fc0685064c6a36f6e5 shows Firefox passing solo.

https://wpt.fyi/results/xslt/transformToFragment.window.html?sha=32c07e34c7&label=pr_head

**Comment by martin-honnen** [comment]
The code I fixed (https://github.com/chromium/chromium/blob/master/third_party/blink/renderer/core/editing/serializers/serialization.cc#L640) is probably in a similar way part of Webkit/Safari, given that is has various copyright notes of Apple. So Safari also failing doesn't surprise me, although I hadn't checked. Does anyone know who authored that requirement note about `transformToFragment` into the HTML W3C and WhatWG spec, have that been Mozilla guys?

**Comment by martin-honnen** [comment]
I have looked at the WebKit sources and in https://github.com/WebKit/webkit/blob/master/Source/WebCore/editing/markup.cpp#L1216 they have, like Chromium, a method `createFragmentForTransformToFragment` that also calls `fragment->parseHTML(sourceString, fakeBody.ptr());` and `fragment->parseXML(sourceString, 0);` without a third argument so the default argument `ParserContentPolicy = AllowScriptingContent` declared in https://github.com/WebKit/webkit/blob/master/Source/WebCore/dom/DocumentFragment.h#L36 kicks in. As Chromium, they also have a `ParserContentPolicy` used by `createContextualFragment` (https://github.com/WebKit/webkit/blob/master/Source/WebCore/dom/Range.cpp#L966), it is declared in https://github.com/WebKit/webkit/blob/master/Source/WebCore/dom/FragmentScriptingPermission.h#L32 and there allows a value `AllowScriptingContentAndDoNotMarkAlreadyStarted`. So I guess WebKit's `createFragmentForTransformToFragment` could be fixed as well by changing the `parseHTML/XML` calls to use `fragment->parseHTML(sourceString, fakeBody.ptr(), AllowScriptingContentAndDoNotMarkAlreadyStarted);` and `fragment->parseXML(sourceString, 0, AllowScriptingContentAndDoNotMarkAlreadyStarted);`.

File: xslt/transformToFragment.window.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
+const cases = {
```
Comment (foolip) [inline xslt/transformToFragment.window.js]: Given no spec, can you make this test tentative?

**Review by foolip — COMMENTED** [review]
Can you add a META.yml for this? Something similar to domxpath/META.yml would do it.

File: xslt/transformToFragment.window.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  );
+  document.body.appendChild(
+    processor.transformToFragment(
+      parser.parseFromString('<x/>', 'application/xml'),
+      document
+    )
+  );
+
+  promise_test(async () => {
+    await Promise.resolve(loaded);
```
Comment (foolip) [inline xslt/transformToFragment.window.js]:
```suggestion
    await loaded;
```

**Review by foolip — APPROVED** [review]
I enjoyed this test :)

**Comment by foolip** [comment]
Hmm, why does https://wpt.fyi/results/?sha=32c07e34c7 not have results for Chrome and Safari?

**Comment by lukebjerring** [comment]
Race condition on pr_base vs pr_head arriving first. I see you already filed https://github.com/web-platform-tests/wpt.fyi/issues/1382

https://wpt.fyi/results/?sha=32c07e34c7&label=pr_head

File: xslt/transformToFragment.window.js (hunk 1 above)
Comment (foolip) [reply xslt/transformToFragment.window.js]: Renamed.

**Comment by foolip** [comment]
@lukebjerring I think only a META.yml is missing now.

**Comment by lukebjerring** [comment]
@foolip - Added a META.yml.

Additionally, I actually manually scraped the IDL and added a test for it:
https://github.com/web-platform-tests/wpt/pull/17607/commits/c8da5386098e66c70b00195ae31d418509d8bb9c#diff-9298fe85f545e6f79dcf5134c331b2bb

As per @annevk pointing out that it's slapped in the [whatwg wiki](https://wiki.whatwg.org/wiki/DOM_XSLTProcessor)

File: interfaces/xslt.idl [hunk 3]
```diff
@@ -0,0 +1,16 @@
+// Content was manually copied July 12 2019
+// Source: https://wiki.whatwg.org/wiki/DOM_XSLTProcessor
+// There is no standard for this API. Only the link above, and notes in
+// the HTML5 standard (https://www.w3.org/TR/html50/scripting-1.html#scriptTagXSLT)
```
Comment (annevk) [inline interfaces/xslt.idl]: https://html.spec.whatwg.org/multipage/scripting.html#scriptTagXSLT

File: xslt/META.yml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+# NOTE: There is no standard for this API, only:
+# https://wiki.whatwg.org/wiki/DOM_XSLTProcessor
+# and the notes in the HTML5 spec, linked.
+spec: https://www.w3.org/TR/html50/scripting-1.html#scriptTagXSLT
```
Comment (annevk) [inline xslt/META.yml]: https://html.spec.whatwg.org/multipage/scripting.html#scriptTagXSLT

File: interfaces/xslt.idl (hunk 3 above)
Comment (lukebjerring) [reply interfaces/xslt.idl]:
```suggestion
// the HTML5 standard (https://html.spec.whatwg.org/multipage/scripting.html#scriptTagXSLT)
```

File: xslt/META.yml (hunk 4 above)
Comment (lukebjerring) [reply xslt/META.yml]:
```suggestion
spec: https://html.spec.whatwg.org/multipage/scripting.html#scriptTagXSLT
```

**Comment by lukebjerring** [comment]
@annevk - FYI, in case you weren't aware, there's a little "suggestion" button when you're authoring a comment that will let you create a tweak to the line of code, which I can then just "Accept".

image

**Comment by annevk** [comment]
I've never found that working (the merging part), but happy to try to remember using that.

**Pull request merged**